cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageRunner.java

```diff
@@ -0,0 +1,291 @@
+                logInfo.setMessage("Could not set job phase to executing, was: " + ep);
+                return;
+            }
+            log.debug(job.getID() + ": QUEUED -> EXECUTING [OK]");
+
+            // Build an iterator of PackageItem from the files named
+            // in the local Job instance
+            Iterator<PackageItem> packageItems = getItems();
+
+            // TarWriter needs an output stream instantiated.
+            initOutputStream(getPackageName());
+            TarWriter writer = new TarWriter(this.outputStreamCounter);
+
+            while (packageItems.hasNext()) {
+                PackageItem pi = packageItems.next();
+                writer.write(pi);
+            }
+
+            // writer is closed in finally clause
+
+            // set final phase, only sync so no results
+            log.debug(job.getID() + ": EXECUTING -> COMPLETED...");
+            ep = jobUpdater.setPhase(job.getID(), ExecutionPhase.EXECUTING, ExecutionPhase.COMPLETED, new Date());
+
+            if (!ExecutionPhase.COMPLETED.equals(ep)) {
+                ep = jobUpdater.getPhase(job.getID());
+                log.debug(job.getID() + ": EXECUTING -> COMPLETED [FAILED], phase was " + ep);
+                logInfo.setSuccess(false);
+                logInfo.setMessage("Could not set job phase to completed.");
+                return;
+            }
+            log.debug(job.getID() + ": EXECUTING -> COMPLETED [OK]");
+
+        } catch (IllegalArgumentException ex) {
+            sendError(ex, ex.getMessage(), 400);
+
+        } catch (UnsupportedOperationException ex) {
+            sendError(ex, "unsupported operation: " + ex.getMessage(), 400);
+
+        } catch (Throwable t) {
+            if (ThrowableUtil.isACause(t, InterruptedException.class)) {
+                try {
+                    ep = jobUpdater.setPhase(job.getID(), ExecutionPhase.QUEUED, ExecutionPhase.EXECUTING, new Date());
+
+                    if (!ExecutionPhase.ABORTED.equals(ep)) {
+                        return; // clean exit of aborted job
+                    }
+
+                } catch (Exception ex2) {
+                    log.error("failed to check job phase after InterruptedException", ex2);
+
+                }
+            }
+            sendError(t, 500);
+        } finally {
+            // Finalize and close writer instance
+            if (writer != null) {
+                try {
+                    writer.close();
+                } catch (Exception ignore) {
+                    log.debug("attempt to close writer when it wasn't open");
+                }
+            }
+        }
+    }
+
+    /**
+     * Initialize syncOutput output stream with the correct content type and disposition,
+     * as provided by the writer class
+     * @param packageName - package name to use
+     * @throws IOException
+     */
+    private void initOutputStream(String packageName) throws IOException {
+        syncOutput.setResponseCode(200);
+
```

Comment:
Either here or up where getPackageName() is called, there needs to be a check for null. I'd normally put it where getPackageName() is called and to keep run() clear I would probably make this method no-arg and call getPackageName() In here.

So, if the subclass returns null, what to do - fail or default value? 
Fail:
```
throw new RuntimeException("BUG: getPackageName() returned null");
```
or default value:
```
packageName = "default-" + job.getID();
```

Reply:
Opted to put a check in initOutputStream, at the least. Left PackageRunner with no default package name.